pkg/skaffold/deploy/helm.go

```diff
@@ -214,8 +213,83 @@ func (h *HelmDeployer) Cleanup(ctx context.Context, out io.Writer) error {
 }
 
 // Render generates the Kubernetes manifests and writes them out
-func (h *HelmDeployer) Render(context.Context, io.Writer, []build.Artifact, []Labeller, string) error {
-	return errors.New("not yet implemented")
+func (h *HelmDeployer) Render(ctx context.Context, out io.Writer, builds []build.Artifact, labellers []Labeller, filepath string) error {
+	hv, err := h.binVer(ctx)
+	if err != nil {
+		return fmt.Errorf("binary version: %w", err)
+	}
+
+	var output io.Writer
+
+	if filepath != "" {
+		f, err := os.OpenFile(filepath, os.O_RDWR|os.O_CREATE, 0666)
+		if err != nil {
+			return fmt.Errorf("opening file for writing manifests: %w", err)
+		}
+		defer f.Close()
+
+		output = f
+	} else {
+		output = out
+	}
+
+	for _, r := range h.Releases {
+		args := []string{"template", r.ChartPath}
+
+		if hv.GTE(helm3Version) {
+			// Helm 3 requires the name to be before the chart path
+			args = append(args[:1], append([]string{r.Name}, args[1:]...)...)
+		} else {
+			args = append(args, "--name", r.Name)
+		}
+
+		for _, vf := range r.ValuesFiles {
+			args = append(args, "--values", vf)
+		}
+
+		params := pairParamsToArtifacts(builds, r.Values)
+
+		for k, v := range params {
+			var value string
+
+			cfg := r.ImageStrategy.HelmImageConfig.HelmConventionConfig
+
+			value, err = imageSetFromConfig(cfg, k, v.Tag)
+			if err != nil {
+				return err
+			}
+
+			args = append(args, "--set-string", value)
+		}
+
+		for key, value := range r.Values {
+			args = append(args, "--set", fmt.Sprintf("%s=%s", key, value))
+		}
+
+		sortedKeys := make([]string, 0, len(r.SetValueTemplates))
+		for k := range r.SetValueTemplates {
+			sortedKeys = append(sortedKeys, k)
+		}
+		sort.Strings(sortedKeys)
+
+		for _, key := range sortedKeys {
+			v, err := util.ExpandEnvTemplate(r.SetValueTemplates[key], nil)
+			if err != nil {
+				return err
+			}
+			args = append(args, "--set", fmt.Sprintf("%s=%s", key, v))
+		}
+
+		if r.Namespace != "" {
+			args = append(args, "--namespace", r.Namespace)
+		}
```

Comment:
is this all that was required for fixing https://github.com/GoogleContainerTools/skaffold/issues/3843?

in general we'd prefer splitting this out into a separate PR so we can track them separately, but I suppose it's fine to keep it here for now.

Reply:
I doubt #3843 is fixed with this MR since I haven't touched any of the Deploy/Cleanup code.

@no1melman are you sure this PR fixes you issue?